DetectorDescription/Core/interface/Store.h

```diff
@@ -9,136 +9,129 @@
-      ~Store();
+      K is the key_type which is used as an index in the storage.
+      It must fulfill all requirements for a key in a sorted associative container.
+      N is the user-friendly name_type, which must be mapped uniquely
+      to the key_type and vice versa. N itself must also fulfill all requirements
+      of a key in a sorted associative container.
+      The reason to provide K is that one might save some memory by compacting
+      common parts of information contained in different instances of N, e.g.
+      if N is a pair<string,string>, the first string being a 'namespace' the 
+      second a 'name' then K could be a pair<int,string> thus compacting the
+      namespace-string to a simple int.
+      K and N must support following unique conversions:
+      - from N to K, thus N(const K &)
+      - from K to N, thus K(const N &)
+  */  
+  template <class N, class I, class K=I>
+    class Store 
+    {
+    public:
+    using name_type = N;
+    using pimpl_type = I;
+    using key_type = K;
+    using Rep_type = rep_type< name_type, pimpl_type >;
+    using prep_type = Rep_type*;
+    using registry_type = std::map< name_type, prep_type>;
+    using iterator = typename registry_type::iterator;
+    
+    auto begin() { return reg_.begin(); }
+    auto end() { return reg_.end(); }
+    auto size() const { return reg_.size(); } 
+    
+    // empty shell or fetch from registry
+    prep_type create(const name_type &);
       
-   protected:
-      std::map<name_type,prep_type> reg_;
-      Store(const Store &);
-      Store & operator=(const Store &);
-      bool readOnly_;
-   };
-   
-   template<class N, class I, class K>
-   typename Store<N,I,K>::prep_type 
-   Store<N,I,K>::create(const name_type & n)
-   {
+    // full new object or replace existing with new one
+    prep_type create(const name_type &, pimpl_type);
+    
+    // clear all objects
+    void clear();
+    
+    // swap moves the registry from this guy to another of the same type
+    void swap ( Store& );
+    
+    bool isDefined(const name_type & n ) const;
+    void setReadOnly(bool b) { readOnly_ = b; }
+    bool readOnly() const { return readOnly_; }
+    
+    Store() : readOnly_(false) { }
+    
+    protected:
+    std::map<name_type,prep_type> reg_;
+    Store(const Store &) = delete;
+    Store & operator=(const Store &) = delete;
+    bool readOnly_;
+    };
+  
+  template<class N, class I, class K>
+    typename Store<N,I,K>::prep_type 
+    Store<N,I,K>::create( const name_type & n )
+    {
       prep_type tmp = nullptr;
-      auto result = reg_.insert(std::make_pair(n,tmp));
-      if (result.second) {
-         if (readOnly_) throw cms::Exception("DetectorDescriptionStore")<<" Store has been locked.  Illegal attempt to add " << n << " to a global store."; 
-         // ELSE     
-         result.first->second = new Rep_type(n,(I)nullptr);
+      auto result = reg_.insert( std::make_pair( n, tmp ));
```

Comment:
A more efficient way is to do 
```
auto result = reg.emplace( n, tmp );
```
this avoids the need to create a temporary pair.